File: leetcode/560-subarray-sum-equals-k/javascript/solution.js

let subarraySum = function(nums, k) {
    let count = 0;
    let sum = 0;
    const map = {};
    map[0] = 1;
    for (let i = 0; i < nums.length; i += 1) {
        sum += nums[i];
        if (map[sum - k]) {
            count += map[sum - k];
        }
        map[sum] = (map[sum] || 0) + 1;
    }
    return count;
};

let subarraySumBruteForce = function(nums, k) {
    let count = 0;
    for (let i = 0; i < nums.length; i += 1) {
        let sum = 0;
        for (let j = i; j < nums.length; j += 1) {
            sum += nums[j];
            if (sum === k) {
                count++;
            }
        }
    }
    return count;
};


console.assert(subarraySumBruteForce([1, 1, 1], 2) === 2, 'Expected 2 but did not get it');
console.assert(subarraySumBruteForce([1, 2, 3], 3) === 2, 'Expected 2 but did not get it');

console.assert(subarraySum([1, 1, 1], 2) === 2, 'Expected 2 but did not get it');
console.assert(subarraySum([1, 2, 3], 3) === 2, 'Expected 2 but did not get it');

File: leetcode/560-subarray-sum-equals-k/python/solution.py
class Solution:
    
    def subarraySumBruteForce(self, nums: list[int], k: int) -> int:
        count = 0
        for i in range(len(nums)):
            sum = 0
            for j in range(i, len(nums)):
                sum += nums[j]
                if sum == k:
                    count += 1
        return count
    
    def subarraySum(self, nums: list[int], k: int) -> int:
        count: int = 0
        sum: int = 0
        prefix_sums: dict = {0: 1}
        for n in nums:
            sum += n
            diff = sum - k
            count += prefix_sums.get(diff, 0)
            prefix_sums[sum] = 1 + prefix_sums.get(sum, 0)
        return count


if __name__ == '__main__':
    sol = Solution()
    assert sol.subarraySumBruteForce([1, 1, 1], 2) == 2, 'Expected 2 but did not get it'
    assert sol.subarraySumBruteForce([1, 2, 3], 3) == 2, 'Expected 2 but did not get it'

    assert sol.subarraySum([1, 1, 1], 2) == 2, 'Expected 2 but did not get it'
    assert sol.subarraySum([1, 2, 3], 3) == 2, 'Expected 2 but did not get it'